### DISCUSSION TITLE
Querying for all instances of a superclass

### DISCUSSION THREAD

**Discussion by johannes-riecken** [go-gorm/gorm#6370] [Q&A]
In our current OpenAPI specification we have several subclasses of one base class, i.e. classes (or schemas) that share properties with another class and also add their own unique properties. We'd like in some cases to query just the specific subclasses (in my example below that's circles and rectangles), and in other cases to query over all superclass instances (shapes in my example), which should then include all instances of subclasses. Would it be a good approach to create a view (using Migrator#CreateView) every time we have such a use case?

```go
package main

import (
	"gorm.io/gorm"
	"gorm.io/driver/postgres"
	"log"
	"fmt"
	"os"
	"gorm.io/gorm/logger"
)

func main() {
	if err := mainAux(); err != nil {
		log.Fatal(err)
	}
}

type Shape struct {
	gorm.Model
	X, Y int
}

type Circle struct {
	gorm.Model
	Shape
	R int
}

type Rectangle struct {
	gorm.Model
	Shape
	W, H int
}

func mainAux() error {
	db, err := gorm.Open(postgres.Open("host=localhost port=5432 dbname=... user=postgres password=..."), &gorm.Config{ })
	if err != nil {
		return err
	}
	db.AutoMigrate(&Shape{}, &Circle{}, &Rectangle{})
	if err = db.Create(&Shape{X: 10, Y: 20}).Error; err != nil {
		return err
	}
	if err = db.Create(&Circle{Shape: Shape{X: 1, Y: 2}, R: 3}).Error; err != nil {
		return err
	}
	if err = db.Create(&Rectangle{Shape: Shape{X: 4, Y: 5}, W: 6, H: 7}).Error; err != nil {
		return err
	}
	shapes := &[]Shape{}
        // I'd like to return all three instances
	if err = db.Find(shapes).Error; err != nil { // doesn't find the subclasses
		return err
	}
	fmt.Println(shapes)
	return nil
}
```

**Comment by roberth1988** [comment]
I would be also interested into an answer for this.

**Comment by RobinKumar5986** [comment]
try this code 
func mainAux() error {
	db, err := gorm.Open(postgres.Open("host=localhost port=5432 dbname=... user=postgres password=..."), &gorm.Config{})
	if err != nil {
		return err
	}
	db.AutoMigrate(&Shape{}, &Circle{}, &Rectangle{})
	if err = db.Create(&Shape{X: 10, Y: 20}).Error; err != nil {
		return err
	}
	if err = db.Create(&Circle{Shape: Shape{X: 1, Y: 2}, R: 3}).Error; err != nil {
		return err
	}
	if err = db.Create(&Rectangle{Shape: Shape{X: 4, Y: 5}, W: 6, H: 7}).Error; err != nil {
		return err
	}
	shapes := []Shape{} // Changed to a slice of Shape, not a pointer to a slice
	if err = db.Find(&shapes).Error; err != nil {
		return err
	}
	fmt.Println(shapes)
	return nil
}

By changing shapes to []Shape{} (a slice of Shape), the db.Find(&shapes) query will fetch all rows from the "shapes" table, including instances of the subclasses (Circle and Rectangle). The result will be stored in the shapes slice, which can then be printed or used further in your code

**Reply by johannes-riecken** [reply]
Hey Robin. Thanks for taking a look at my problem. I still don't get the subclasses returned (gorm v1.25.1). Also your code seems to be equivalent to mine, because you use `db.Find(&shapes)` to still pass the pointer to the slice to the Find method.
